nixos/modules/services/misc/matrix-conduit.nix

```diff
@@ -0,0 +1,168 @@
+        example = "pkgs.matrix-conduit";
+        description = ''
+          Package of the conduit matrix server to use.
+        '';
+      };
+
+      settings = mkOption {
+        type = types.submodule {
+          freeformType = format.type;
+          options = {
+            global.server_name = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "example.com";
+              description = "The server_name is the name of this server. It is used as a suffix for user # and room ids.";
+            };
+            global.port = mkOption {
+              type = types.port;
+              default = 6167;
+              description = "The port Conduit will be running on. You need to set up a reverse proxy in your web server (e.g. apache or nginx), so all requests to /_matrix on port 443 and 8448 will be forwarded to the Conduit instance running on this port";
+            };
+            global.max_request_size = mkOption {
+              type = types.ints.positive;
+              default = 20000000;
+              description = "Max request size in bytes. Don't forget to also change it in the proxy.";
+            };
+            global.allow_registration = mkOption {
+              type = types.bool;
+              default = false;
+              description = "Whether new users can register on this server.";
+            };
+            global.allow_encryption = mkOption {
+              type = types.bool;
+              default = false;
+              description = "Whether new encrypted rooms can be created. Note: existing rooms will continue to work.";
+            };
+            global.allow_federation = mkOption {
+              type = types.bool;
+              default = false;
+              description = ''
+                Whether this server federates with other servers.
+                Upstream documentation contains the following warning:
+                You should not join Matrix rooms without asking the admins first.
+                We do not know whether Conduit is safe for general use yet, so you
+                should assume there is some chance that it breaks rooms permanently
+                for all participating users. We are not aware of such a bug today,
+                but we would like to do more testing.
+              '';
+            };
+            global.allow_jaeger = mkOption {
+              type = types.bool;
+              default = false;
+              description = "Whether jaeger to support monitoring and troubleshooting through jaeger is enabled on this server.";
+            };
+            global.trusted_servers = mkOption {
+              type = types.listOf types.str;
+              default = [ "matrix.org" ];
+              description = "Servers trusted with signing server keys.";
+            };
+
+            global.address = mkOption {
+              type = types.str;
+              default = "127.0.0.1";
```

Comment:
Use `::1` instead. It is 2021 we should support IPv6 adoption (even on localhost) :)

Reply:
Upstream defaults... Also some upstream complexities: conduit **needs** to have the address in the form of `::1`, nginx needs to have the address as part of an URL as `[::1]` for proxying. Is there some library function that can transform one representation in the other without affecting hostnames or ipv4?

Reply:
Not to my knowledge. Usually people do something like `isIPv6 = ip: builtins.length (lib.splitString ":" ip) > 2;` (taken from the nat module). You could probably just use `lib.hasInfix ":" addr` to decide that..